### PR TITLE
Validate cluster UUID when joining Zen1 cluster

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we fail to join a Zen2 cluster if the cluster UUID does not match our
own, but we do not perform the same validation when joining a Zen1 cluster.
This means that a Zen2 node will pass join validation and be added to a Zen1
cluster but will reject all cluster states from the master.

Relates #37775

### CODE REVIEW

**Comment by DaveCTurner** [comment]
I will forward-port the relevant part of this test to `master` in a followup.

File: server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java [hunk 1]
```diff
@@ -153,6 +153,13 @@ public JoinHelper(Settings settings, AllocationService allocationService, Master
         transportService.registerRequestHandler(MembershipAction.DISCOVERY_JOIN_VALIDATE_ACTION_NAME,
             ValidateJoinRequest::new, ThreadPool.Names.GENERIC,
             (request, channel, task) -> {
+                final ClusterState localState = currentStateSupplier.get();
+                if (localState.metaData().clusterUUIDCommitted() &&
+                    localState.metaData().clusterUUID().equals(request.getState().metaData().clusterUUID()) == false) {
+                    throw new CoordinationStateRejectedException("join validation on cluster state" +
+                        " with a different cluster uuid " + request.getState().metaData().clusterUUID() +
+                        " than local cluster uuid " + localState.metaData().clusterUUID() + ", rejecting");
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java]: Perhaps add a zen1 marker to the message to aid in debugging?

File: server/src/test/java/org/elasticsearch/cluster/coordination/JoinHelperTests.java [hunk 2]
```diff
@@ -131,4 +139,59 @@ public void testFailedJoinAttemptLogLevel() {
...
+            deterministicTaskQueue.getThreadPool(), TransportService.NOOP_TRANSPORT_INTERCEPTOR,
+            x -> localNode, null, Collections.emptySet());
+        new JoinHelper(Settings.EMPTY, null, null, transportService, () -> 0L, () -> localClusterState,
+            (joinRequest, joinCallback) -> { throw new AssertionError(); }, startJoinRequest -> { throw new AssertionError(); },
+            Collections.emptyList()); // registers request handler
+        transportService.start();
+        transportService.acceptIncomingRequests();
+
+        {
+            final ClusterState otherClusterState = ClusterState.builder(ClusterName.DEFAULT).metaData(MetaData.builder()
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/coordination/JoinHelperTests.java]: The two blocks for zen1 and zen2 are identical except for the action name, I would prefer to extract the method to make this clear.

**Review by henningandersen — APPROVED** [review]
LGTM.

Thanks @DaveCTurner , I added two optional comments to consider.

File: server/src/test/java/org/elasticsearch/cluster/coordination/JoinHelperTests.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/coordination/JoinHelperTests.java]: Ok, done in 1dfd207.

**Pull request merged**